README.md
# Flysystem Stream Wrapper

[![Author](https://img.shields.io/badge/author-@m2mtech-blue.svg?style=flat-square)](http://www.m2m.at)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)

---

This package provides a stream wrapper for Flysystem V2 & V3.

## Flysystem V1

If you're looking for Flysystem 1.x support, check out the [twistor/flysystem-stream-wrapper](https://github.com/twistor/flysystem-stream-wrapper).

## Installation

```bash
composer require m2mtech/flysystem-stream-wrapper
```

## Usage

```php
use League\Flysystem\Filesystem;
use League\Flysystem\Local\LocalFilesystemAdapter;
use M2MTech\FlysystemStreamWrapper\FlysystemStreamWrapper;

$filesystem = new Filesystem(new LocalFilesystemAdapter('/some/path'));
FlysystemStreamWrapper::register('fly', $filesystem);

file_put_contents('fly://filename.txt', $content);
mkdir('fly://happy_thoughts');

FlysystemStreamWrapper::unregister('fly');
```

The stream wrapper implements [`symfony/lock`](https://symfony.com/doc/current/components/lock.html) due to Flysystem V2 not supporting locking. By default, file locking using `/tmp` is used, but you can adjust this through configuration:

```php
FlysystemStreamWrapper::register('fly', $filesystem, [
    FlysystemStreamWrapper::LOCK_STORE => 'flock:///tmp',
    FlysystemStreamWrapper::LOCK_TTL => 300,
]);
```

### Handling Visibility Issues

Some adaptors might throw exceptions when dealing with visibility. If you encounter such issues, configure the stream wrapper to bypass them:

```php
FlysystemStreamWrapper::register('fly', $filesystem, [
    FlysystemStreamWrapper::IGNORE_VISIBILITY_ERRORS => true,
]);
```

### Addressing Directory Issues (`file_exists` / `is_dir`)

Some adaptors might not return dates for the last modified attribute for directories. In such cases, you can enable emulation to achieve the desired behavior:

```php
FlysystemStreamWrapper::register('fly', $filesystem, [
    FlysystemStreamWrapper::EMULATE_DIRECTORY_LAST_MODIFIED => true,
]);
```

### Dealing with `is_readable` / `is_writable`

Some filesystem functions depend on the `uid` and `gid` of the user executing PHP. Since a reliable cross-platform method to derive these values isn't available, the wrapper attempts to estimate them. If this fails, set them manually:

```php
FlysystemStreamWrapper::register('fly', $filesystem, [
    FlysystemStreamWrapper::UID => 1000,
    FlysystemStreamWrapper::GID => 1000,
]);
```

Alternatively, access the parameters for [`PortableVisibilityConverter`](https://flysystem.thephpleague.com/docs/usage/unix-visibility/) directly:

```php
FlysystemStreamWrapper::register('fly', $filesystem, [
    FlysystemStreamWrapper::VISIBILITY_FILE_PUBLIC => 0644,
    FlysystemStreamWrapper::VISIBILITY_FILE_PRIVATE => 0600,
    FlysystemStreamWrapper::VISIBILITY_DIRECTORY_PUBLIC => 0755,
    FlysystemStreamWrapper::VISIBILITY_DIRECTORY_PRIVATE => 0700,
    FlysystemStreamWrapper::VISIBILITY_DEFAULT_FOR_DIRECTORIES => Visibility::PRIVATE,
]);
```

## Testing

This package was developed using PHP 7.4 and has been tested for compatibility with PHP versions 7.2 through 8.3.

To test:

- With PHP installed:
```bash
composer test
```

- Inside a Docker environment for PHP 7.4:
```bash
docker compose run php74 composer test
```

**Note**: PHPUnit v10, used from PHP 8.1 onwards, requires a different config file:
```bash
docker compose run php81 composer test10
```

## Changelog

For information on recent changes, refer to the [CHANGELOG](CHANGELOG.md).

## Contributing

For contribution guidelines, see [CONTRIBUTING](.github/CONTRIBUTING.md).

## Security Vulnerabilities

If you discover any security vulnerabilities, please follow [our security policy](../../security/policy) for reporting.

## Credits

- This package was inspired by [twistor/flysystem-stream-wrapper](https://github.com/twistor/flysystem-stream-wrapper). Many thanks to [Chris Leppanen](https://github.com/twistor).
- [All Contributors](../../contributors)

## License

Licensed under the MIT License. See the [License File](LICENSE.md) for more details.


CHANGELOG.md
# Changelog

## [1.4.0] - 2023-12-18
- compatibility with symfony 7.x & php 8.3
- allow ecs 12.x
- added `EMULATE_DIRECTORY_LAST_MODIFIED`, thanks to @das-peter

## [1.3.3] - 2023-03-27
- fixed tests for phpunit v10

## [1.3.2] - 2023-03-21
- allow `bool` in `FlysystemStreamWrapper::register` configuration
- fixed phpunit deprecations  
- updated (c) dates

## [1.3.1] - 2022-08-12
- intercept TypeError when adapter returns null for visibility

## [1.3.0] - 2022-07-21
- adjusted uid & gid retrieval 
- added manual uid & gid setup
- allow to access parameters of [`PortableVisibilityConverter`](https://flysystem.thephpleague.com/docs/usage/unix-visibility/)
- fixed typo in const `IGNORE_VISIBILITY_ERRORS`

## [1.2.1] - 2022-06-10
- added intl for dev to fix ecs dependency
- fix for resources closed by flysystem prematurely

## [1.2.0] - 2022-06-04
- set current user uid & gid in stats

## [1.1.0] - 2022-05-05
- allow for adaptors closing the file handle themselves
- fixed broken directory detection caused by changed MimeType detection of [Flysystem 3.0.16](https://github.com/thephpleague/flysystem/compare/3.0.15...3.0.16)
- allow to ignore visibility errors

## [1.0.3] - 2022-01-21
- allow flysystem v3

## [1.0.2] - 2022-01-10
- replace section in [composer.json](composer.json)

## [1.0.1] - 2021-12-29
- compatibility with symfony 6.x
- skipped unit tests of locking for php8.1 till opis/closure is available in version 4.x (only used for testing)

## [1.0.0] - 2021-11-01
- Initial release 

<!---
## [Unreleased]
### Changed
- ...
--->

[Unreleased]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.3.3...HEAD
[1.4.0]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.3.3...v1.4.0
[1.3.3]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.3.2...v1.3.3
[1.3.2]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.3.1...v1.3.2
[1.3.1]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.3.0...v1.3.1
[1.3.0]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.2.1...v1.3.0
[1.2.1]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.0.3...v1.1.0
[1.0.3]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/m2mtech/flysystem-stream-wrapper/releases/tag/v1.0.0
<!---
[Unreleased]: https://github.com/m2mtech/flysystem-stream-wrapper/compare/v1.3.2...HEAD
--->


src/FlysystemStreamWrapper.php
<?php
/*
 * This file is part of the flysystem-stream-wrapper package.
 *
 * (c) 2021-2023 m2m server software gmbh <tech@m2m.at>
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace M2MTech\FlysystemStreamWrapper;

use League\Flysystem\FilesystemOperator;
use League\Flysystem\Visibility;
use M2MTech\FlysystemStreamWrapper\Flysystem\Helper\UserGuesser;
use M2MTech\FlysystemStreamWrapper\Flysystem\StreamWrapper;

final class FlysystemStreamWrapper
{
    public const LOCK_STORE = 'lock_store';
    public const LOCK_TTL = 'lock_ttl';

    /** @deprecated 2.0.0 use FlysystemStreamWrapper::IGNORE_VISIBILITY_ERRORS instead */
    public const IGNORE_VISIBILITY_ERROS = 'ignore_visibility_errors';

    public const IGNORE_VISIBILITY_ERRORS = 'ignore_visibility_errors';

    public const EMULATE_DIRECTORY_LAST_MODIFIED = 'emulate_directory_last_modified';

    public const UID = 'uid';
    public const GID = 'gid';

    public const VISIBILITY_FILE_PUBLIC = 'visibility_file_public';
    public const VISIBILITY_FILE_PRIVATE = 'visibility_file_private';
    public const VISIBILITY_DIRECTORY_PUBLIC = 'visibility_directory_public';
    public const VISIBILITY_DIRECTORY_PRIVATE = 'visibility_directory_private';
    public const VISIBILITY_DEFAULT_FOR_DIRECTORIES = 'visibility_default_for_directories';

    public const DEFAULT_CONFIGURATION = [
        self::LOCK_STORE => 'flock:///tmp',
        self::LOCK_TTL => 300,

        self::IGNORE_VISIBILITY_ERRORS => false,
        self::EMULATE_DIRECTORY_LAST_MODIFIED => false,

        self::UID => null,
        self::GID => null,

        self::VISIBILITY_FILE_PUBLIC => 0644,
        self::VISIBILITY_FILE_PRIVATE => 0600,
        self::VISIBILITY_DIRECTORY_PUBLIC => 0755,
        self::VISIBILITY_DIRECTORY_PRIVATE => 0700,
        self::VISIBILITY_DEFAULT_FOR_DIRECTORIES => Visibility::PRIVATE,
    ];

    /** @var array <string, FilesystemOperator> */
    public static $filesystems = [];

    /** @var array <string, array<string, int|string|bool|null>> */
    public static $config = [];

    /** @param array<string, bool|int|string> $configuration */
    public static function register(
        string $protocol,
        FilesystemOperator $filesystem,
        array $configuration = [],
        int $flags = 0
    ): bool {
        if (self::streamWrapperExists($protocol)) {
            return false;
        }

        self::$config[$protocol] = array_merge(self::DEFAULT_CONFIGURATION, $configuration);
        self::$filesystems[$protocol] = $filesystem;

        if (null === self::$config[$protocol][self::UID]) {
            self::$config[$protocol][self::UID] = UserGuesser::getUID();
        }

        if (null === self::$config[$protocol][self::GID]) {
            self::$config[$protocol][self::GID] = UserGuesser::getGID();
        }

        return stream_wrapper_register($protocol, StreamWrapper::class, $flags);
    }

    public static function unregister(string $protocol): bool
    {
        if (!self::streamWrapperExists($protocol)) {
            return false;
        }

        unset(self::$config[$protocol], self::$filesystems[$protocol]);

        return stream_wrapper_unregister($protocol);
    }

    public static function unregisterAll(): void
    {
        foreach (self::getRegisteredProtocols() as $protocol) {
            self::unregister($protocol);
        }
    }

    /** @return array<int, string> */
    public static function getRegisteredProtocols(): array
    {
        return array_keys(self::$filesystems);
    }

    public static function streamWrapperExists(string $protocol): bool
    {
        return in_array($protocol, stream_get_wrappers(), true);
    }
}
